metals/src/main/scala/scala/meta/internal/metals/Diagnostics.scala

```diff
@@ -0,0 +1,66 @@
+package scala.meta.internal.metals
+
+import ch.epfl.scala.bsp4j
+import ch.epfl.scala.bsp4j.CompileReport
+import java.util
+import java.util.Collections
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ConcurrentLinkedQueue
+import org.eclipse.lsp4j.Diagnostic
+import org.eclipse.lsp4j.PublishDiagnosticsParams
+import org.eclipse.lsp4j.services.LanguageClient
+import scala.collection.concurrent.TrieMap
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+
+class Diagnostics(bsp: BuildTargets, languageClient: LanguageClient) {
+  private val diagnostics =
+    TrieMap.empty[AbsolutePath, util.Queue[Diagnostic]]
+  private val hasPublishedDiagnostics =
+    Collections
+      .newSetFromMap(new ConcurrentHashMap[AbsolutePath, java.lang.Boolean]())
+      .asScala
+
+  def onBuildPublishDiagnostics(
+      params: bsp4j.PublishDiagnosticsParams
+  ): Unit = {
+    if (params.getDiagnostics.isEmpty) {
+      () // Do nothing.
+    } else {
+      val path = params.getUri.toAbsolutePath
+      val queue =
+        diagnostics.getOrElseUpdate(
+          path,
+          new ConcurrentLinkedQueue[Diagnostic]()
+        )
+      params.getDiagnostics.forEach { buildDiagnostic =>
+        queue.add(buildDiagnostic.toLSP)
+      }
+      languageClient.publishDiagnostics(
```

Comment:
This is exponential in number of errors, let's document that problem and open issue to report only at boundaries of documents.